data/src/main/java/org/apache/iceberg/data/DeleteFilter.java

```diff
@@ -96,6 +98,29 @@ public Schema requiredSchema() {
     return requiredSchema;
   }
 
+  protected int deleteMarkerIndex() {
+    if (deleteMarkerIndex != null) {
+      return deleteMarkerIndex;
+    }
+
+    int index = 0;
+    for (Types.NestedField field : requiredSchema().columns()) {
+      if (field.fieldId() != MetadataColumns.IS_DELETED.fieldId()) {
+        index = index + 1;
+      } else {
+        break;
+      }
+    }
+
+    deleteMarkerIndex = index;
+
+    return deleteMarkerIndex;
+  }
+
+  protected abstract Consumer<T> deleteMarker();
```

Comment:
How about introducing a new interface named `Setter` to set the `is_deleted` flag (which is similar to the org.apache.iceberg.Accessor)  so that we could have a good abstraction to hide the delete marker logic: 

```java
  interface Setter<T> extends Serializable {
    T set(T reuse);
  }
```

Reply:
I can try to make this better abstraction in the following PR, this PR contains too many changes now.  I think we will have some following-up minor changes and optimizations. Does that sound ok to you?